### PR TITLE
Update ROOT formula options

### PR DESCRIPTION
**Pull request by henryiii**
This makes several cleanups and updates to better support ROOT 6.16.00:

* ~~Dropped need to use thisroot.sh to use PyROOT (normal ROOT already works without it). Removed caveat about thisroot (it is considered bad to recommend thisroot by the ROOT team)~~: This can't be done without a hacky patch at the moment. The ROOT team is targeting a new version of PyROOT for ~6.20 with this fixed.
* ROOT expects the CMake standard to be set on the command line (this instructs CLING to support this level of C++) - by default, you only get C++11. This increases it to the maximum supported on the system (C++14 or C++17 depending on macOS version). In ROOT 6.18, this will change to `-DCMAKE_CXX_STANDARD=` instead of these flags.
* An extra test was added to make sure you can build with ROOT.
* You can now build with `-I/usr/local/include -I/usr/local/include/root` again. This is not included in above test, since `root-config` (correctly) chooses the more specific Cellar include directories. Will be fixed in ROOT 6.16.02.

### CODE REVIEW

**Comment by henryiii** [comment]
I may be running into the bug reported here: https://trac.macports.org/ticket/57007

Don't merge until I have looked into it (even if the tests start passing).

**Comment by henryiii** [comment]
Quick question: Is there a recommended way to do a multithreaded build? Something like `-j${CORES}`? This would be 2x faster locally if I could use two cores, and I'd guess that CircleCI might have more than one core (Travis has 2).

I have one possible fix applied, and now testing something the ROOT team recommended as an alternative.

**Comment by henryiii** [comment]
Nevermind, found it in the Boost formula, `-j#{ENV.make_jobs}`

**Comment by henryiii** [comment]
Okay, so here are the possible solutions:

1. Replace `<>` with `""` in `RConfig.h`. This keeps the compiler from reincluding the same file, thinking it is `ROOT/RConfig.h` (since it's in `root`).
2. Delete `RConfig.h` and symlink `ROOT/RConfig.h` to it. This is what macPorts does.
3. Use `-gnuinstall=OFF`. This avoids the `/include/root` structure entirely, and instead puts all 1K or so includes directly in `/include`. This means a user should not include `/usr/local/include/root`, since that's now `/usr/local/include/ROOT`. I don't think this change should be made just to fix this problem.
4. There's a more complex rename solution we are [working on](https://github.com/root-project/root/pull/3361), but it's not an easy patch for homebrew. This hopefully will come in ROOT 16.02 and whatever solution we do now can be removed then.

So, I'm in favor of 1) or 2). I've implemented 1) (partially because I know how to do that easily in Homebrew). Does anyone else have an opinion?

**Comment by henryiii** [comment]
Ready to merge unless someone prefers option 2.

**Comment by mistydemeo** [comment]
I think option 1 sounds like the cleanest solution to me. Thanks for documenting it so clearly!

File: Formula/root.rb
Comment (mistydemeo) [inline Formula/root.rb]: If this is making a change to how C++ support is being compiled, is it worth bumping the revision for it?

Comment (henryiii) [reply Formula/root.rb]: We can. It now matches the maximum allowed on your system. I believe this mostly affects people using the C++ interpreter; it will now report a higher C++ value if asked. It also allows ROOT to use several features in the stdlib that otherwise would use a custom backport, but that shouldn't affect users.

Do I just add a `_0` to the version?

Comment (mistydemeo) [reply Formula/root.rb]: Just add `revision 1` underneath the version!

Comment (henryiii) [reply Formula/root.rb]: Done!

**Comment by henryiii** [comment]
@SMillerDev, I think the do-not-merge label can be removed now, it's ready for review.

**Comment by henryiii** [comment]
Fix 4 was merged into the ROOT sources, and will be part of 6.16.02. Fix 2 is still a good interim solution until 6.16.02 comes out, IMO, since fix 4 is a much harder patch that changes 50 or so files.

File: Formula/root.rb
Comment (amadio) [inline Formula/root.rb]: `-Dtbb=ON` does not exist. `-Dimt=ON` is really the correct option.

File: Formula/root.rb
Comment (amadio) [reply Formula/root.rb]: Are there any packages that depend on ROOT in Homebrew (for instance, rivet, geant4-vmc, etc)? If yes, then you might need to make sure they depend on ROOT with the same C++ standard, or they will likely fail to compile.

**Review by amadio — CHANGES_REQUESTED** [review]
Looks good to me except for the `-Dtbb=ON`, which doesn't exist. As for the issue with linking PyROOT directly to Python, we may have to patch the homebrew recipe, or go back to the previous setup and add the extra directory to `PYTHONPATH` in the default `bashrc` of Homebrew.

**Comment by henryiii** [comment]
Is there a way to temporarily disable the check for files linked to libPython in site-packages? This is a hardcoded problem with ROOT for the past (many) years. It is just as present in the current version; the problem just escapes the automated test because you have to run `thisroot.sh` to change the PYTHONPATH, they are not linked into site-packages.

I do have a patch that does fix this that I use in the Conda package, but it's a bit ugly and hacky. I can try to come up with a nicer patch, probably won't be done very quickly though.

File: Formula/root.rb
Comment (henryiii) [reply Formula/root.rb]: Okay, fixed this for next time I push. Thanks!

**Comment by henryiii** [comment]
PS: The problem is ROOT both allows Python to load libPyROOT.so, and ROOT can load it too. Unless all users add `gSystem->Load('libpython3.7m")` or something similar before using `TPython` in ROOT, they will get an unresolved symbols error if it is not linked.

**Comment by henryiii** [comment]
Still planning the improved patch to ROOT. If it goes well, the target date is to get it done before the 21st of March (since I will be giving a talk over the conda-forge ROOT package then). Might slip past that though.

That patch should fix the main failing check here (as well as make ROOT behave more nicely with Homebrew Python).

File: Formula/root.rb
Comment (fxcoudert) [inline Formula/root.rb]:
```suggestion
    system "#{bin}/root", "-b", "-l", "-q", "-e", "'gSystem->LoadAllLibraries(); 0'"
```

**Review by fxcoudert — CHANGES_REQUESTED** [review]

File: Formula/root.rb
Comment (fxcoudert) [reply Formula/root.rb]: This is failing with:

```
==> root -b -l -q -e 'gSystem->LoadAllLibraries(); 0'

ROOT_cli_0:1:1: warning: multi-character character constant [-Wmultichar]
'gSystem->LoadAllLibraries(); 0'
^
ROOT_cli_0:1:1: warning: character constant too long for its type
(int) 691740720
Error: root: failed
```

I believe the single quotes should be removed.

**Comment by henryiii** [comment]
This should be ready to merge, unless you want to wait for ROOT 6.16.02, which was "coming soon" a month ago.

**Review by fxcoudert — APPROVED** [review]

**Comment by fxcoudert** [comment]
@BrewTestBot test this please

**Comment by fxcoudert** [comment]
Thank you very much @henryiii

**Comment by henryiii** [comment]
Thanks @fxcoudert! When 6.16.02 comes out I'll prepare a PR for that one.

**Pull request closed without merging**
